proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,190 @@
+last slot is always last_voted_slot, first slot is last_voted_slot-81000.
+
+### Aggregate, repair, and replay the slots in LastVotedForkSlots
+
+Aggregate the slots in received LastVotedForkSlots messages, whenever some slot
+has enough stake to be optimistically confirmed and it's missing locally, start
+the repair process for this slot.
+
+We calculate "enough" stake as follows. When there are 80% validators joining
+the restart, assuming 5% restarted validators can make mistakes in voting, any
+block with more than 67% - 5% - (100-80)% = 42% could potentially be
+optimistically confirmed before the restart. If there are 85% validators in the
+restart, then any block with more than 67% - 5% - (100-85)% = 47% could be
+optimistically confirmed before the restart.
+
+### Gossip current heaviest fork
+
+After receiving LastVotedForkSlots from 80% of the validators and reparing
+slots with "enough" stake, replay all blocks and pick the heaviest fork as
+follows:
+
+1. Pick block and update root for all blocks with more than 67% votes
+
+2. If a picked block has more than one children, check if the votes on the
+heaviest child is over threshold:
+
+2.1 If vote_on_child + stake_on_validators_not_in_restart >= 62%, pick child.
+For example, if 80% validators are in restart, child has 42% votes, then
+42 + (100-80) = 62%, pick child. 62% is chosen instead of 67% because 5%
+could make the wrong votes.
+
+2.2 Otherwise stop traversing the tree and use last picked block.
+
+After deciding heaviest block, Gossip
+Heaviest(X, Hash(X), received_heaviest_stake) out, where X is the latest picked
+block. We also send out stake of received Heaviest messages so that we can
+proceed to next step when enough validators are ready.
+
+The fields of the Heaviest message is:
+
+- `slot`: slot of the picked block.
+- `hash`: bank hash of the picked block.
+- `received`: total of stakes of the validators it received Heaviest messages
+from.
+
+### Proceed to restart if everything looks okay, halt otherwise
+
+If things go well, all 80% of the validators should find the same heaviest
+fork. But we are only sending slots instead of bank hashes in
+LastVotedForkSlots, so it's possible that a duplicate block can make the
+cluster unable to reach consensus. If at least 2/3 of the people agree on one
+slot, they should proceed to restart from this slot. Otherwise validators
+should halt and send alerts for human attention.
+
+Also, there might be 5% of the validators not sending Heaviest, so we only
+require that 75% of the people received 75% of the Heaviest messages and they
+all agree on one block and hash.
+
+So after a validator sees that 75% of the validators received 75% of the votes,
+wait for 10 more minutes so that the message it sent out have propagated, then
+do the following:
+
+- Issue a hard fork at the highest oc slot and change shred version in Gossip.
+- Execute the current --wait-for-supermajority logic and wait for 75%.
```

Comment:
Also wondering if we should create a new snapshot after the hard fork. If the validator is restarted for some reason in the the current --wait-for-supermajority logic, it'd be nice for it to resume from the new snapshot instead of needing to restart from 0

Reply:
The original proposal is to generate a snapshot, but there were concerns that we need to change logic which says we only generate snapshot at the root. I personally feel generating snapshot is a safer choice.

Reply:
Ya in general for sure. This is the specific restart case though where we also deliberately hard fork at the highest oc slot and by doing so it effectively becomes a root as it'll never be rolled back

Reply:
Okay, I'll add that back. What is the reason we only generate snapshot at the root? Is it because other slots might be rolled back so there is no point generating at non-roots?

Reply:
Yeah exactly. A snapshot currently holds but a single `Bank` (one day I'd love to see that changed to hold a full `BankForks`), so when a node starts from a snapshot there's literally no ancestor banks loaded